### PR TITLE
fix: prevent OOM in Automodel recipes

### LINKED ISSUE #3842: Track nightly OOMs after the Automodel r0.6.0 upgrade
## Summary

PR #3498 upgrades Automodel to r0.6.0. Older Automodel versions loaded model modules in BF16 even when an FP32 load dtype was requested. This was fixed in v0.6.0, but the increased memory usage causes these nightly tests to OOM:

- `llm_dpo_nanov3_30B3AB_1n8g_fsdp8ep8_automodel`
- `llm_grpo_nanov3_30BA3B_2n8g_fsdp2_tq_mooncake`
- `llm_grpo_nanov3_30BA3B_2n8g_fsdp2`
- `vlm_grpo_nemotron_omni_30ba3b_clevr_1n8g_automodel_ep8_v1`
- `grpo-nemotron3-super-120BA12B-16n8g-automodel-ep8`

Track fixes for these cases after #3498 merges.

### PR DESCRIPTION
**Pull request by jQizhang**
Updates the affected Automodel recipes and test suites with memory-safe configurations.

Fixes https://github.com/NVIDIA-NeMo/RL/issues/3842

### CODE REVIEW

File: tests/test_suites/llm/dpo-nanov3-30B3AB-1n8g-fsdp8ep8-automodel.v2.sh [hunk 1]
```diff
@@ -34,11 +34,10 @@ uv run tests/json_dump_tb_logs.py $LOG_DIR --output_path $JSON_METRICS
...
     uv run tests/check_metrics.py $JSON_METRICS \
-        'data["train/loss"]["1"] < 0.69316' \
-        'data["train/loss"]["11"] < 0.53' \
-        'data["train/preference_loss"]["1"] > 0.69314' \
-        'data["train/preference_loss"]["1"] < 0.69316' \
-        'data["train/preference_loss"]["11"] < 0.53' \
+        'data["train/loss"]["1"] < 0.694' \
+        'data["train/loss"]["11"] < 0.57' \
+        'data["train/preference_loss"]["1"] < 0.694' \
+        'data["train/preference_loss"]["11"] < 0.57' \
```
Comment (jQizhang) [inline tests/test_suites/llm/dpo-nanov3-30B3AB-1n8g-fsdp8ep8-automodel.v2.sh:40]: [`tests/test_suites/llm/dpo-nanov3-30B3AB-1n8g-fsdp8ep8-automodel.v2.sh:37-40`](https://github.com/NVIDIA-NeMo/RL/blob/f208ac862d84899ec3ed5e7c359c45e3f31387d9/tests/test_suites/llm/dpo-nanov3-30B3AB-1n8g-fsdp8ep8-automodel.v2.sh#L37-L40)

**1 action item.**

The only two config deltas in this recipe ([`env_vars.PYTORCH_CUDA_ALLOC_CONF`](https://github.com/NVIDIA-NeMo/RL/blob/f208ac862d84899ec3ed5e7c359c45e3f31387d9/examples/configs/recipes/llm/dpo-nanov3-30B3AB-1n8g-fsdp8ep8-automodel.v2.yaml#L23-L24), [`dtensor_cfg.moe_parallelizer.reshard_after_forward`](https://github.com/NVIDIA-NeMo/RL/blob/f208ac862d84899ec3ed5e7c359c45e3f31387d9/examples/configs/recipes/llm/dpo-nanov3-30B3AB-1n8g-fsdp8ep8-automodel.v2.yaml#L27-L28)) are both numerics-inert by contract: PyTorch FSDP2's `fully_shard(reshard_after_forward=...)` only controls all-gather scheduling (no reduction, not floating-point-order sensitive), and Automodel's `apply_fsdp()` in [`parallelizer.py`](https://github.com/NVIDIA-NeMo/Automodel/blob/1814c6c93a66b9d59d254960ef6a99a64249b671/nemo_automodel/components/moe/parallelizer.py) passes it straight through — the only conditional keyed on this flag (for MTP blocks) just picks which bool value to pass, with no dtype-casting or numeric branching. `PYTORCH_CUDA_ALLOC_CONF` is allocator-only.

Yet the checked thresholds moved:
- `train/loss["11"]`: `< 0.53` → `< 0.57` (~7.5% relative)
- `train/preference_loss["1"]`: the **lower bound `> 0.69314` was removed entirely** — a step-1, pre-optimizer-step sanity check (asserting near-chance-level ln(2) loss before any gradient update) that should be config-invariant regardless of memory/scheduling settings. This exact bound was introduced with a tight 0.00002-wide window and has been carried forward since.

Since both disclosed config deltas are provably numerics-inert, the drift isn't explained by anything in this diff — which raises the question of whether something else changed numerically (e.g. an Automodel version pin bump bundled into this PR) that isn't mentioned in the PR description at all. That makes the missing quantitative evidence more concerning, not less: it's not just "please confirm this is fine," it's "what actually produced this drift?"

**Action:** Please share the before/after `train/loss` and `train/preference_loss` curves (wandb/tensorboard) that motivated these threshold changes, confirm whether anything besides the two disclosed settings changed numerically (e.g. a dependency/pin bump), and explain why the step-1 `preference_loss` lower bound was dropped rather than restored/widened with a measured margin.

_rl-expert, bug-finder, test-agent — independently converged; CONFIRMED unanimously by devil-advocate after rl-expert traced `parallelizer.py` directly_

File: tests/test_suites/llm/common-tq.env [hunk 2]
```diff
@@ -9,14 +9,19 @@
...
+else
+    echo "[ERROR] Script must end in -tq_<backend>[.vN]: $TQ_EXP_NAME" >&2
+    return 1
 fi
 
-if [[ "$TQ_EXP_NAME" != *-tq_simple && "$TQ_EXP_NAME" != *-tq_mooncake ]]; then
-    echo "[ERROR] Script must end in -tq_simple or -tq_mooncake: $TQ_EXP_NAME" >&2
+if [[ ! "$BASE_RECIPE" =~ ^(vlm_)?(grpo|dapo|prorlv2)- ]]; then
+    echo "[ERROR] TQ coverage is only for (vlm_)grpo/dapo/prorlv2 recipes; got: $BASE_RECIPE" >&2
     exit 1
```
Comment (jQizhang) [inline tests/test_suites/llm/common-tq.env:26]: [`tests/test_suites/llm/common-tq.env:21-27`](https://github.com/NVIDIA-NeMo/RL/blob/f208ac862d84899ec3ed5e7c359c45e3f31387d9/tests/test_suites/llm/common-tq.env#L21-L27)

**1 action item.**

PR-introduced inconsistency: the new malformed-name branch (line 21) uses `return 1` — correct for a sourced-only file (every caller uses `source "$SCRIPT_DIR/common-tq.env"`, never executes it directly) — while the unchanged branch at line 26 still uses `exit 1`, which would kill the parent script's process outright rather than just failing the sourcing operation. The latent smell is actually on the old `exit 1` side; this PR only fixed it in one of the two branches.

**Action:** For consistency, match the newer (more correct) pattern:
```suggestion
    return 1
fi
```

_bug-finder; framing confirmed by devil-advocate over an earlier "pre-existing, not introduced" read_

File: tests/test_suites/llm/common-tq.env [hunk 3]
```diff
@@ -9,14 +9,19 @@
...
 
-if [[ ! "$BASE_RECIPE" =~ ^(vlm_)?(grpo|dapo|prorlv2)- ]]; then
-    echo "[ERROR] TQ coverage is only for (vlm_)grpo/dapo/prorlv2 recipes; got: $BASE_RECIPE" >&2
-    exit 1
+# The canonical .vN suffix follows all modifiers. Keep accepting the legacy
+# <base>.vN-tq_<backend> spelling while deriving the matching base recipe.
+if [[ "$TQ_EXP_NAME" =~ ^(.+)-tq_(simple|mooncake)(\.v[0-9]+)$ ]]; then
+    BASE_RECIPE="${BASH_REMATCH[1]}${BASH_REMATCH[3]}"
+elif [[ "$TQ_EXP_NAME" =~ ^(.+)-tq_(simple|mooncake)$ ]]; then
+    BASE_RECIPE="${BASH_REMATCH[1]}"
```
Comment (jQizhang) [inline tests/test_suites/llm/common-tq.env:18]: [`tests/test_suites/llm/common-tq.env`](https://github.com/NVIDIA-NeMo/RL/blob/f208ac862d84899ec3ed5e7c359c45e3f31387d9/tests/test_suites/llm/common-tq.env) — regex-parsing logic for `BASE_RECIPE` derivation

**No action needed** — offered as an optional follow-up, not a blocking ask.

test-agent wrote and verified a derivation-logic test (covers all 24 real `-tq_*` wrappers in the repo plus synthetic edge cases; confirmed it catches an injected regression). Checked how it would actually run in CI: `tests/test_suites/llm/*.sh` only execute when listed in [`nightly.txt`](https://github.com/NVIDIA-NeMo/RL/blob/f208ac862d84899ec3ed5e7c359c45e3f31387d9/tests/test_suites/nightly.txt) (GPU training-run drivers only), and there's no shellcheck/test_suites lint job in `.github/workflows` — so a standalone `.sh` file here would be invisible to CI. This parsing logic currently has zero test coverage anywhere in the repo.

**Suggestion:** rather than a loose `tests/test_suites/llm/*.sh` script, add this as a `pytest` test under `tests/unit/` that subprocess-invokes the bash derivation logic — that gets picked up automatically by the existing `CI:L0` unit-test run, with no nightly-list or shellcheck plumbing needed.

_test-agent, downgraded and refined by devil-advocate_

**Review by jQizhang — COMMENTED** [review]
## PR #4034 review — fix: prevent OOM in Automodel recipes

Reviewed by a 5-agent team (rl-expert, test-agent, bug-finder, design-reviewer, devil-advocate). No merge conflicts. Confirmed non-trivial: genuine memory-tuning across 4 recipe YAMLs plus a coordinated `.vN` rename pass and the `common-tq.env` fix needed to keep TQ coverage matching post-rename.

No design-level surface (diff is entirely YAML/docs/bash — no `.py` files touched) — design-reviewer's LGTM confirmed by devil-advocate.

Linter (`pre-commit run --all-files`, including `minimize-check llm recipes`): all hooks passed.

1 finding needs author input before merge (test-rigor regression with no stated justification), 2 low-severity suggestions below.

_Generated by Claude Code_

File: tests/test_suites/llm/dpo-nanov3-30B3AB-1n8g-fsdp8ep8-automodel.v2.sh (hunk 1 above)
Comment (sharonyu-115) [inline tests/test_suites/llm/dpo-nanov3-30B3AB-1n8g-fsdp8ep8-automodel.v2.sh:40]: Shall we at the meantime adding gen_kl_error check to the script?

File: examples/configs/recipes/llm/grpo-nanov3-30BA3B-1n8g-fsdp2.v2.yaml [hunk 4]
```diff
@@ -12,12 +12,22 @@ policy:
   train_micro_batch_size: 1
   logprob_batch_size: 1
   max_total_sequence_length: 2048
+  optimizer:
```
Comment (sharonyu-115) [inline examples/configs/recipes/llm/grpo-nanov3-30BA3B-1n8g-fsdp2.v2.yaml:15]: Overall the changes look good to me. One question is how much perf impact the changes introduce to the recipes? Would it be a concern of hitting the `total_gpu_hours` allowed for nightly test?

**Review by sharonyu-115 — CHANGES_REQUESTED** [review]

File: tests/test_suites/llm/dpo-nanov3-30B3AB-1n8g-fsdp8ep8-automodel.v2.sh (hunk 1 above)
Comment (jQizhang) [reply tests/test_suites/llm/dpo-nanov3-30B3AB-1n8g-fsdp8ep8-automodel.v2.sh:40]: Hi @sharonyu-115, this is a DPO experiment script, so gen_kl_error isn’t included in the metrics.

File: examples/configs/recipes/llm/grpo-nanov3-30BA3B-1n8g-fsdp2.v2.yaml (hunk 4 above)
Comment (jQizhang) [reply examples/configs/recipes/llm/grpo-nanov3-30BA3B-1n8g-fsdp2.v2.yaml:15]: I’ve summarized the runtimes for these scripts in the table below. They’re all within the max-gpu-time limit, so they shouldn’t go over the total_gpu_hours budget.

YAML File | Actual Runtime | Maximum Time Limit | Exceeded?
-- | -- | -- | --
dpo-nanov3-30B3AB-1n8g-fsdp8ep8-automodel.v2.yaml | Approximately 00:06:05 | 30 minutes | No
grpo-nanov3-30BA3B-1n8g-fsdp2.v2.yaml | 00:23:30 | 50 minutes | No
grpo-nanov3-30BA3B-1n8g-fsdp2-tq_mooncake.v2.yaml | 00:26:18 | 45 minutes | No
vlm_grpo-nemotron-omni-30ba3b-clevr-1n8g-automodel-ep8.v2.yaml | 01:40:44 | 120 minutes | No
grpo-nemotron3-super-120BA12B-16n8g-automodel-ep8.v2.yaml | 00:23:03 | 120 minutes | No*

File: tests/test_suites/llm/dpo-nanov3-30B3AB-1n8g-fsdp8ep8-automodel.v2.sh (hunk 1 above)
Comment (sharonyu-115) [reply tests/test_suites/llm/dpo-nanov3-30B3AB-1n8g-fsdp8ep8-automodel.v2.sh:40]: Got it. Thank you!

**Review by sharonyu-115 — APPROVED** [review]
LGTM. Thanks!
